### PR TITLE
ChannelStore migration Part 2

### PR DESCRIPTION
**Pull request by rvillablanca**
#### Summary

Second part of migration of `ChannelStore` to return plain errors.

#### Ticket Link
 
Fixes: #15125 

#### Jira Ticket:

https://mattermost.atlassian.net/browse/MM-27318

### CODE REVIEW

**Comment by rvillablanca** [comment]
@agnivade could you tell me what I can do to resolve the `ci/circleci: check-deps` check ?

**Comment by agnivade** [comment]
It's not code-related. @jupenur should be able to help.

**Comment by jupenur** [comment]
@rvillablanca I updated the bad cache and retriggered the workflow; it should be working now.

**Comment by rvillablanca** [comment]
Thanks @jupenur

File: store/sqlstore/channel_store.go [hunk 1]
```diff
@@ -1999,43 +1979,43 @@ func (s SqlChannelStore) GetGuestCount(channelId string, allowFromCache bool) (i
...
+	query, args, err := builder.ToSql()
 	if err != nil {
-		return model.NewAppError("SqlChannelStore.RemoveMember", "store.sql_channel.remove_member.app_error", nil, "channel_id="+channelId+", "+err.Error(), http.StatusInternalServerError)
+		return errors.Wrap(err, "channel_tosql")
 	}
-	_, err = s.GetMaster().Exec(sql, args...)
+	_, err = s.GetMaster().Exec(query, args...)
 	if err != nil {
-		return model.NewAppError("SqlChannelStore.RemoveMember", "store.sql_channel.remove_member.app_error", nil, "channel_id="+channelId+", "+err.Error(), http.StatusInternalServerError)
+		return errors.Wrap(err, "failed to Delete ChannelMembers")
```
Comment (agnivade) [inline store/sqlstore/channel_store.go:1998]: Any reason `Delete` with a capital D?

File: store/sqlstore/channel_store.go [hunk 2]
```diff
@@ -1999,43 +1979,43 @@ func (s SqlChannelStore) GetGuestCount(channelId string, allowFromCache bool) (i
...
 		}).ToSql()
 	if err != nil {
-		return model.NewAppError("SqlChannelStore.RemoveMember", "store.sql_channel.remove_member.app_error", nil, "channel_id="+channelId+", "+err.Error(), http.StatusInternalServerError)
+		return errors.Wrap(err, "channel_tosql")
 	}
-	_, err = s.GetMaster().Exec(sql, args...)
+	_, err = s.GetMaster().Exec(query, args...)
 	if err != nil {
-		return model.NewAppError("SqlChannelStore.RemoveMember", "store.sql_channel.remove_member.app_error", nil, "channel_id="+channelId+", "+err.Error(), http.StatusInternalServerError)
+		return errors.Wrap(err, "failed to Delete SidebarChannels")
```
Comment (agnivade) [inline store/sqlstore/channel_store.go:2013]: Same as above

File: store/sqlstore/channel_store.go [hunk 3]
```diff
@@ -2101,21 +2081,12 @@ func (s SqlChannelStore) UpdateLastViewedAt(channelIds []string, userId string)
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to find ChannelMembers data with userId=%s and channelId in %v", userId, channelIds)
+	}
 
-		return nil, model.NewAppError("SqlChannelStore.UpdateLastViewedAt",
-			"store.sql_channel.update_last_viewed_at.app_error",
-			nil,
-			"channel_ids="+strings.Join(channelIds, ",")+", user_id="+userId+", "+extra,
-			status)
+	if len(lastPostAtTimes) == 0 {
```
Comment (agnivade) [inline store/sqlstore/channel_store.go:2088]: Nice simplification :)

**Review by agnivade — COMMENTED** [review]
Tremendous piece of work! Appreciate the effort Rodrigo!

File: app/channel.go [hunk 4]
```diff
@@ -119,7 +120,23 @@ func (a *App) JoinDefaultChannels(teamId string, user *model.User, shouldBeAdmin
 
 	}
 
-	return err
+	if nErr != nil {
+		var appErr *model.AppError
+		var cErr *store.ErrConflict
+		switch {
+		case errors.As(err, &cErr):
+			switch cErr.Resource {
```
[the remark this answers is not in the document]
Comment (rvillablanca) [reply app/channel.go:128]: yeah, thinks of copy-pase.

File: app/channel.go [hunk 5]
```diff
@@ -119,7 +120,23 @@ func (a *App) JoinDefaultChannels(teamId string, user *model.User, shouldBeAdmin
...
+			case "ChannelMembers":
+				return model.NewAppError("JoinDefaultChannels", "app.channel.save_member.exists.app_error", nil, cErr.Error(), http.StatusBadRequest)
+			}
+		case errors.As(err, &appErr):
+			return appErr
+		default:
+			return model.NewAppError("JoinDefaultChannels", "app.channel.create_direct_channel.internal_error", nil, err.Error(), http.StatusInternalServerError)
```
[the remark this answers is not in the document]
Comment (rvillablanca) [reply app/channel.go:135]: :+1:

File: app/channel.go [hunk 6]
```diff
@@ -1038,9 +1081,18 @@ func (a *App) UpdateChannelMemberRoles(channelId string, userId string, newRoles
...
-		return nil, err
+	member, nErr := a.Srv().Store.Channel().UpdateMember(member)
+	if nErr != nil {
+		var appErr *model.AppError
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(nErr, &appErr):
+			return nil, appErr
+		case errors.As(nErr, &nfErr):
+			return nil, model.NewAppError("UpdateChannelMemberRoles", store.MISSING_CHANNEL_MEMBER_ERROR, nil, nfErr.Error(), http.StatusNotFound)
```
[the remark this answers is not in the document]
Comment (rvillablanca) [reply app/channel.go:1092]: Yep we can move it to another place as it is not being used anymore in the store package. When do you suggests? something like `constants.go` in the app package or in in this file `channel.go`.

File: app/channel.go [hunk 7]
```diff
@@ -2158,7 +2266,13 @@ func (a *App) SetActiveChannel(userId string, channelId string) *model.AppError
 
 func (a *App) UpdateChannelLastViewedAt(channelIds []string, userId string) *model.AppError {
 	if _, err := a.Srv().Store.Channel().UpdateLastViewedAt(channelIds, userId); err != nil {
-		return err
+		var invErr *store.ErrInvalidInput
+		switch {
+		case errors.As(err, &invErr):
+			return model.NewAppError("UpdateChannelLastViewedAt", "app.channel.update_last_viewed_at.app_error", nil, err.Error(), http.StatusBadRequest)
```
[the remark this answers is not in the document]
Comment (rvillablanca) [reply app/channel.go:2271]: Fixed

File: store/constants.go [hunk 8]
```diff
@@ -4,7 +4,7 @@
 package store
 
 const (
-	MISSING_CHANNEL_MEMBER_ERROR = "store.sql_channel.get_member.missing.app_error"
+	MISSING_CHANNEL_MEMBER_ERROR = "app.channel.get_member.missing.app_error"
```
[the remark this answers is not in the document]
Comment (rvillablanca) [reply store/constants.go:7]: Ok I'm going to move it to a new file `constants.go` in the app package

File: store/sqlstore/channel_store.go [hunk 9]
```diff
@@ -1730,12 +1717,16 @@ func (s SqlChannelStore) GetMemberForPost(postId string, userId string) (*model.
 		AND
 			Posts.Id = :PostId`
 	if err := s.GetReplica().SelectOne(&dbMember, query, map[string]interface{}{"UserId": userId, "PostId": postId}); err != nil {
-		return nil, model.NewAppError("SqlChannelStore.GetMemberForPost", "store.sql_channel.get_member_for_post.app_error", nil, "postId="+postId+", err="+err.Error(), http.StatusInternalServerError)
+		if err == sql.ErrNoRows {
```
[the remark this answers is not in the document]
Comment (rvillablanca) [reply store/sqlstore/channel_store.go:1720]: got it

File: store/sqlstore/channel_store.go (hunk 1 above)
Comment (rvillablanca) [reply store/sqlstore/channel_store.go:1998]: just a typo

File: store/sqlstore/channel_store.go (hunk 2 above)
Comment (rvillablanca) [reply store/sqlstore/channel_store.go:2013]: :man_facepalming:

File: store/sqlstore/channel_store.go (hunk 3 above)
Comment (rvillablanca) [reply store/sqlstore/channel_store.go:2088]: :)

**Comment by rvillablanca** [comment]
@agnivade This is ok with the suggestions/bugs so can you make the review again? Thanks!

**Review by agnivade — APPROVED** [review]

**Comment by rvillablanca** [comment]
@jespino Friendly reminder to review this (and others if possible :slightly_smiling_face: ) PR

**Review by jespino — APPROVED** [review]
LGTM. Thanks @rvillablanca and sorry for the delay.

**Comment by rvillablanca** [comment]
@agnivade @jespino who should I assign the QA review :thinking: ?

**Comment by agnivade** [comment]
@lindalumitchell - Can be qa-deferred.

**Review by lindalumitchell — APPROVED** [review]

**Pull request merged**